plugins/Synchro/SynchroSynth.cpp

```diff
@@ -0,0 +1,389 @@
+	m_modulatorDetune.loadSettings(thisElement, "modulatorDetune");
+	m_modulatorDrive.loadSettings(thisElement, "modulatorDrive");
+	m_modulatorSync.loadSettings(thisElement, "modulatorSync");
+	m_modulatorChop.loadSettings(thisElement, "modulatorChop");
+}
+
+QString SynchroSynth::nodeName() const
+{
+	return synchro_plugin_descriptor.name;
+}
+
+void SynchroSynth::playNote(NotePlayHandle * n, sampleFrame * working_buffer)
+{
+	if (n->totalFramesPlayed() == 0 || n->m_pluginData == NULL)
+	{
+		n->m_pluginData = new SynchroNote(n);
+	}
+
+	const fpp_t frames = n->framesLeftForCurrentPeriod();
+	const f_cnt_t offset = n->noteOffset();
+
+	SynchroNote * ps = static_cast<SynchroNote *>(n->m_pluginData);
+	for(fpp_t frame = offset; frame < frames + offset; ++frame)
+	{
+		sampleFrame outputSample = {0, 0};
+		sampleFrame tempSample = {0, 0};
+		SynchroOscillatorSettings carrier = { m_carrierDetune.value(),
+			m_carrierDrive.value(), m_carrierSync.value(),
+			m_carrierChop.value() };
+		SynchroOscillatorSettings modulator = { m_modulatorDetune.value(),
+			m_modulatorDrive.value(), m_modulatorSync.value(),
+			m_modulatorChop.value() };
+
+		//Oversampling using linear interpolation
+		for (int i = 0; i < SYNCHRO_OVERSAMPLE; ++i)
+		{
+			ps->nextStringSample(tempSample,
+				Engine::mixer()->processingSampleRate()
+				* SYNCHRO_OVERSAMPLE, m_modulationStrength.value(),
+				m_modulation.value(),m_harmonics.value(), carrier,
+				modulator);
+			outputSample[0] += tempSample[0];
+			outputSample[1] += tempSample[1];
+		}
+
+		for(ch_cnt_t chnl = 0; chnl < DEFAULT_CHANNELS; ++chnl)
+		{
+			working_buffer[frame][chnl] = outputSample[chnl]
+				/ SYNCHRO_OVERSAMPLE;
+		}
+	}
+	applyRelease(working_buffer, n);
+	instrumentTrack()->processAudioBuffer(working_buffer, frames + offset, n);
+}
+
+void SynchroSynth::deleteNotePluginData(NotePlayHandle * n)
+{
+	delete static_cast<SynchroNote *>(n->m_pluginData);
+}
+
+PluginView * SynchroSynth::instantiateView(QWidget * parent)
+{
+	return new SynchroSynthView(this, parent);
+}
+
+void SynchroSynth::carrierChanged()
+{
+	for (int i = 0; i < SYNCHRO_GRAPH_SAMPLES; ++i)
+	{
+		float phase = (float)i / (float)SYNCHRO_GRAPH_SAMPLES;
+		m_carrierGraph.setSampleAt(i, SynchroWaveform(phase * F_2PI,
+			m_carrierDrive.value(), m_carrierSync.value(),
+			m_carrierChop.value(), 0));
+	}
+	generalChanged();
+}
+
+void SynchroSynth::modulatorChanged()
+{
+	for (int i = 0; i < SYNCHRO_GRAPH_SAMPLES; ++i)
+	{
+		float phase = (float)i / (float)SYNCHRO_GRAPH_SAMPLES;
+		m_modulatorGraph.setSampleAt(i, SynchroWaveform(phase * F_2PI,
+			m_modulatorDrive.value(), m_modulatorSync.value(),
+			m_modulatorChop.value(), m_harmonics.value()));
+	}
+	generalChanged();
+}
+
+void SynchroSynth::generalChanged()
+{
+	for (int i = 0; i < SYNCHRO_GRAPH_SAMPLES; ++i)
+	{
+		int octaveDiff = m_carrierDetune.value() - m_modulatorDetune.value();
+		float pitchDifference = powf(2, octaveDiff);
```

Comment:
These values can't be `const` at all, since they can (and are) changed by knob input. `octaveDiff` is only used once and can be inlined into `pitchDifference` for readability, though.

I may be able to factor them out of the loop, but only if I were to opt against making the oscillator octave controls sample-exact.

Reply:
`const` only applies for the duration of the function's execution, so that wouldn't stop it getting a different value for the knob the next time the function was called. Since the variables referred to aren't modified during the function (once they've been factored out of the loop, that is), `const` is applicable.

Bear in mind that the graphs only show a snapshot of the waveform at each point in time. The code here will run on a scale of nanoseconds, but a sample of audio is on the scale of microseconds and a whole buffer of audio is on the scale of milliseconds. The samples generated here are along an "imaginary" time axis, not real time, and are generated on a different thread to the audio, so the question of how sample-exact automation corresponds to samples on the graph is pretty much meaningless. Your best bet is just to pull the value of the knobs at the start, and go off that.

Reply:
This is my bad; I thought this was in reference the audio synthesis code at first. This makes much more sense now that I see it's the graph.

Reply:
It seems like this one is still not addressed.